freezer/src/freezer_files.rs

```diff
@@ -0,0 +1,366 @@
+use crate::internal_error;
+use ckb_error::Error;
+use std::collections::BTreeMap;
+use std::convert::TryInto;
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::io::{Seek, SeekFrom};
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+const MAX_FILE_SIZE: u64 = 2 * 1_000 * 1_000 * 1_000;
+const INDEX_FILE_NAME: &str = "INDEX";
+const INDEX_ENTRY_SIZE: u64 = 12;
+
+pub type FileId = u32;
+
+pub(crate) struct Head {
+    pub(crate) file: File,
+    pub(crate) bytes: u64,
+}
+
+impl Head {
```

Comment:
This struct is not safe after `write` returns error. The field `bytes` may be out of sync. Because freezer is running in the background and continue on errors (see [shared.rs](https://github.com/nervosnetwork/ckb/pull/2119/files#r443977331)), all the structs must still in the consistent states or can recover automatically on errors.

The issue here is that `write_all` may fail and written only partial data to file. That partially written bytes are not counted in `bytes`.